### PR TITLE
User management & permission management rewrite

### LINKED ISSUE #59: Add user management (master task)
Right now all permission modification needs to happen through the database. It's now time that the basics are down that we change that. Honestly, get creative with how you want to handle the form, but do realize a user can have multiple db entries in the permissions table (wiki being the only proper distinction), and that there are even duplicate entries right now (which will need a scripted sql patch of sorts to go with the pull request for this issue.

Tasklist:
- [ ] Make sure multiple entries in the permissions table are handled
- [ ] Create a mysql script of sorts to remove current duplicates (should be a separate issue)
- [ ] Prevent duplicate user verifications which cause duplicate inserts into permissions table (#88)
- [ ] Properly handle who can manage users and which permissions they can add
- [ ] Make sure a reason for user changes is obtained on submission
- [ ] Add a logs table entry for the change
- [ ] Show all previous userrights changes log (should be a separate issue)
- [ ] Only allow changing of non-wiki based userrights: Developer, tooladmin & privacy (Developers can add all; tooladmin can add privacy for wikis in which they are a tool admin.
- [ ] Show wiki-based rights read-only
- [ ] Create laravel job for verifying current onwiki permissions on a regular basis  (should be a separate issue - HIGH PRIORITY, before this issue is dealt with) (#88)

There is likely more, so i'll modify this as we go through it and add/replace tasks as necessary.

### LINKED ISSUE #90: Add useful links to navbar to logged in users / users with permissions
The current navbar is empty. We could add some useful links such as appeal list or template/user/ban management.

This is much easier when #106 is merged as that allows using `@can('viewAny', \App\Template::class)` styled statements in blade files.

### LINKED ISSUE #130: Page for own profile
We need a page that allows users to:
- [ ] <s>Reset their password & forgot password (using wikimail if any email is sent)</s>
- [ ] View the wikis and permissions on each wiki they have been assigned. 

This is seperate from user management as it's tooladmin only there

### LINKED ISSUE #164: Label the logout button Logout
Currently the logout button displays as the user's username:
image

When I clicked on it the first time, I thought it would go to a profile/settings page instead of being logged out.

It should say logout or perhaps be similar to ACC:
image

### PR DESCRIPTION
**Pull request by supertassu**
This PR is a rewrite of the user and permission management systems. It effectively closes #59. Bullet points and their status from #59:

- [x] Make sure multiple entries in the permissions table are handled
- [x] Create a mysql script of sorts to remove current duplicates (should be a separate issue)
- [x] Prevent duplicate user verifications which cause duplicate inserts into permissions table (handled in #88)
- [x] Properly handle who can manage users and which permissions they can add
- [x] Make sure a reason for user changes is obtained on submission
- [x] Add a logs table entry for the change
- [x] Show all previous userrights changes log (should be a separate issue)
- [x] Only allow changing of non-wiki based userrights: Developer, tooladmin & privacy (Developers can add all; tooladmin can add privacy for wikis in which they are a tool admin.
- [x] Show wiki-based rights read-only
- [x] Create laravel job for verifying current onwiki permissions on a regular basis  (should be a separate issue - HIGH PRIORITY, before this issue is dealt with) (handled in #88)

### CODE REVIEW

**Comment by supertassu** [comment]
This is ready for review, however please review #88 before reviewing this one as this one has a couple of commented-out features that depend on 88. As always, please ask if you don't understand something.

**Comment by dqwiki** [comment]
In running my usability tests before running through code review, I thought I would lay out 4 thoughts:

- The MySQL script to remove duplicates can definitely be a different issue, but needs to be put as such
- There is no ability to add users to global ('*') groups. 
- Also, I'm not sure how far reverifying wiki permissions goes if it tries to add new wikis the user isn't already a part of (which it should)
- Users should be able to manually refresh their own permissions with the wiki.

I'm not saying any of this needs to be in this PR, but just writing this as a usability of this feature report. So effectively, I wouldn't say this closes #59 but it definitely handles a lot of it.

File: app/Policies/Admin/UserPolicy.php
Comment (dqwiki) [inline app/Policies/Admin/UserPolicy.php]: Even though I don't see the implementation of this, no one should be able to create a user as it's meant to be OAuth managed

**Review by dqwiki — COMMENTED** [review]
Commenting for now as I only have 16/23 files reviewed

**Comment by supertassu** [comment]
>   * There is no ability to add users to global ('*') groups.

It's possible if the user has ever been in a global group (= there is a row in permissions table). I'll try to figure out a way to add them for.
>   * Also, I'm not sure how far reverifying wiki permissions goes if it tries to add new wikis the user isn't already a part of (which it should)

It checks in all wikis configured in `config/wikis.php`


>   * Users should be able to manually refresh their own permissions with the wiki.

Not a bad idea. {{done}}

**Comment by supertassu** [comment]
> The MySQL script to remove duplicates can definitely be a different issue, but needs to be put as such

Created artisan command `utrs-maintenance:remove-duplicate-permissions` for that

**Review by dqwiki — APPROVED** [review]
I added a dry run command that should be looked over before I merge this.

I'm marking this as approved. I'll let you decide if you want to include the dealing with global group in this PR or not.

File: app/Console/Commands/RemoveDuplicatePermissionsCommand.php
Comment (dqwiki) [inline app/Console/Commands/RemoveDuplicatePermissionsCommand.php:11]: Disagree with the removal of the '?'. Dry is an optional command and the braces don't make it optional.
https://laravel.com/docs/6.x/artisan#arguments

**Review by dqwiki — CHANGES_REQUESTED** [review]

File: app/Console/Commands/RemoveDuplicatePermissionsCommand.php
Comment (supertassu) [reply app/Console/Commands/RemoveDuplicatePermissionsCommand.php:11]: '?' is not needed for flags:

```
C:\Users\taavi\sites\utrs>php artisan utrs-maintenance:remove-duplicate-permissions

 You are about to run a live database change. Are you sure? (Did you test it?) (yes/no) [no]:
 > no

Operation aborted.

C:\Users\taavi\sites\utrs>php artisan utrs-maintenance:remove-duplicate-permissions --dry
No duplicates found.
```

**Comment by supertassu** [comment]
> I'm marking this as approved. I'll let you decide if you want to include the dealing with global group in this PR or not.

Implemented.

File: app/Console/Commands/RemoveDuplicatePermissionsCommand.php
Comment (dqwiki) [reply app/Console/Commands/RemoveDuplicatePermissionsCommand.php:11]: Oh I guess it's seen as an option in the next section on that laravel doc page which doesn't require it...interesting.

**Review by dqwiki — APPROVED** [review]

**Pull request merged**
